src/persist-client/src/impl/gc.rs

```diff
@@ -195,6 +196,7 @@ impl GarbageCollector {
         // concurrently, but this requires a bunch of Arc cloning, so wait to
         // see if it's worth it.
         for key in deleteable_blobs {
+            let key = PartialBlobKey(key).complete(&req.shard_id);
```

Comment:
oooof this is a really good find, and also quite painful.

I wonder if `persist-client` should have its own type like `KeyedBlob` that delegates to a `Blob` implementation internally, but wraps the calls to require `BlobKey`. Could help delineate the structure that `persist-client` is trying to layer on top of the more generic blob store, right now it's hard to keep track of what portion of the key we have... could look something like:

```rust
pub struct KeyedBlob {
    blob: Arc<dyn Blob + Send + Sync>,
}

impl KeyedBlob {
    async fn get(&self, key: &BlobKey) -> Result<Option<Vec<u8>>, ExternalError> {
        self.blob.get(key).await
    }

    async fn list_keys_and_metadata(
        &self,
        key_prefix: &BlobKeyPrefix,
        f: &mut (dyn FnMut(BlobMetadata) + Send + Sync),
    ) -> Result<(), ExternalError> {
        self.blob.list_keys_and_metadata(key_prefix, f).await
    }

    async fn set(
        &self,
        key: &BlobKey,
        value: Bytes,
        atomic: Atomicity,
    ) -> Result<(), ExternalError> {
        self.blob.set(&key, value, atomic).await
    }

    async fn delete(&self, key: &BlobKey) -> Result<(), ExternalError> {
        self.blob.delete(key)
    }
}
```

Reply:
if we introduce the blob cache again, it would be an ideal place to put this